### ISSUE TITLE
Support for Registering External Functions and Derivatives

### ISSUE THREAD

**Issue by ysyecust** [symengine/symengine#2054]
Hello SymEngine Team,

I hope you're doing well! I have a question regarding the support for external function registration and derivative handling in SymEngine.

In the Julia Symbolics library, there is a feature that allows users to register external functions and specify how their derivatives should be handled, as described in the following link: [Julia Symbolics - Registering Functions](https://symbolics.juliasymbolics.org/v6.13/manual/functions/#Registering-Functions).

I would like to ask if SymEngine supports a similar feature, where users can register custom functions (including specifying how their derivatives should be computed). If so, could you please provide some guidance or examples on how to achieve this?

Thank you for your time and for the excellent work on this library! Looking forward to your response.

**Comment by bjodah** [comment]
When you say "computed" do you mean evaluated as floating point, or something else?

**Comment by ysyecust** [comment]
> When you say "computed" do you mean evaluated as floating point, or something else?

yes I mean evaluated as floating point , the registration function like this 
``` cpp
double foo (double x,double y){
return x^2+y^2;
}

```
I want to use the function foo in symbolic system , and when give the value of "x, y" I can evaluate the "foo" like 
```cpp
expr = x+foo(x,y);
```

**Comment by bjodah** [comment]
You should be able to subclass the LambdaRealDoubleVisitor already today I think. For LLVMDoubleVisitor, I think you might need to patch it to accept foreign functions, but shouldn't be much more than registering the signature (and maybe inject symbols into the jit machinery, I don't know for sure without taking a closer look).

**Comment by ysyecust** [comment]
> You should be able to subclass the LambdaRealDoubleVisitor already today I think. For LLVMDoubleVisitor, I think you might need to patch it to accept foreign functions, but shouldn't be much more than registering the signature (and maybe inject symbols into the jit machinery, I don't know for sure without taking a closer look).

To add, my problem is that my custom function is usually a black box function with no expression
``` cpp 
double foo (double x,double y){
return blackbox(x,y);
}

```

**Comment by shifengqu** [comment]
I think `FunctionWrapper` can work for defining external functions. 

```
class KCalc : public FunctionWrapper {
   public:
    KCalc(RCP<const Basic> arg) : FunctionWrapper("KCalc", arg) {}

    RCP<const Number> eval() const override {
        return real_double(custom_func(eval_double(*get_vec()[0])));
    }

    RCP<const Basic> create(const vec_basic &v) const override {
        if (eq(*zero, *v[0])) {
            return zero;
        } else {
            return make_rcp<KCalc>(v[0]);
        }
    }

    RCP<const Basic> diff_impl(const RCP<const Symbol> &x) const override {
        return mul(custom_derivative(get_vec()[0]), get_vec()[0]->diff(x));
    }
};

// Custom function definitions
double custom_func(double x) {
    return x * x;  // Example: f(x) = x^2
}

double custom_derivative(double x) {
    return 2 * x;  // Example: f'(x) = 2x
}
```

However, I want to raise a point regarding the differentiation of functions with multiple variables.

In my case, if I have a function with two or more variables, such as foo(x, y), the derivative can differ depending on which variable I differentiate with respect to. The current implementation of diff_impl seems to return a single expression.

Is there a way to modify diff_impl to handle multiple variables appropriately, allowing it to return different expressions based on the variable specified for differentiation?

**Comment by isuruf** [comment]
> Is there a way to modify diff_impl to handle multiple variables appropriately, allowing it to return different expressions based on the variable specified for differentiation?

I'm not sure I understand. `diff_impl` is supposed to return the total derivative with respect to the symbol given.
